pkg/utils/clusterclient/clusterclient.go

```diff
@@ -0,0 +1,123 @@
+package clusterclient
+
+import (
+	"fmt"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+	clusterv1alpha1 "kubesphere.io/kubesphere/pkg/apis/cluster/v1alpha1"
+	typed_v1alpha1 "kubesphere.io/kubesphere/pkg/client/clientset/versioned/typed/application/v1alpha1"
+	clusterinformer "kubesphere.io/kubesphere/pkg/client/informers/externalversions/cluster/v1alpha1"
+	listers_v1alpha1 "kubesphere.io/kubesphere/pkg/client/listers/application/v1alpha1"
+	"net/url"
+	"sync"
+)
+
+var (
+	ClusterNotExistsFormat = "cluster %s not exists"
+)
+
+type clusterClients struct {
```

Comment:
shares lots of code with `pkg/apiserver/dispatch`, we need to make reusable

Reply:
https://github.com/kubesphere/kubesphere/pull/3334/commits/05f54f4827b6c5920181628c8ec01136073d55c8